spectree/plugins/falcon_plugin.py

```diff
@@ -209,6 +209,17 @@
                     (entry.dict() if isinstance(entry, BaseModel) else entry)
                     for entry in model
                 ]
+            elif expect_model and is_root_model(expect_model):
+                if not isinstance(model, expect_model):
+                    # Make it possible to return an instance of the model __root__ type
+                    # (i.e. not the root model itself).
+                    try:
+                        model = expect_model(__root__=model)
+                    except ValidationError:
```

Comment:
Not 100% sure I understand your question - are you asking if we should skip this part entirely when skip_validation (provided by user) is True?

Reply:
I think `skip_validation==True` should skip all the pydantic conversion/validation since users should handle the response by themselves when they enable it. Otherwise, we should raise a ValidationError exception here. What do you think?

Reply:
> I think skip_validation==True should skip all the pydantic conversion/validation since users should handle the response by themselves when they enable it

💯

Then for all plugins, we should simply skip the entire validation block. For example for `flask_plugin.py` (line 211):

```python
if resp:
```

Would become:

```python
if resp and not skip validation:
```

What do you think?

> Otherwise, we should raise a ValidationError exception here. What do you think?

I don't 100% agree here. The user could return a serialized value that conforms to the provided model/schema. In that case, we want to rely on the final validation to confirm whether the returned value is valid/not. I'm referring to the following logic (lines 244-251 in flask_plugin.py):

```python
            if model and not skip_validation:
                try:
                    model.parse_obj(response.get_json())
                except ValidationError as err:
                    resp_validation_error = err
                    response = make_response(
                        jsonify({"message": "response validation error"}), 500
                    )
```

Reply:
1. yeah, we can skip the entire validation block
2. Can you provide an example? I feel that these two validations are almost the same.

BTW, if it's validated by the 1st one successfully, then shall we set the `skip_validation` to skip the next one?

Reply:
> Can you provide an example? I feel that these two validations are almost the same.

### Example 1 (simple model)

```python
class MyResponse(BaseModel):
    id: int
    name: str

@validate(resp=Response(HTTP_200=MyResponse))
def my_endpoint()
    return {"id": 1, "name": "user1"}
```

Here, we return a payload that conforms to the specified model/schema. The existing logic will not fail validation, event though we did not return a MyResponse instance. This is thanks to logic between lines 245-251 in the flask plugin (but it's similar for other plugins).

If possible, I would like to keep that behaviour for root models.

### Example 2 (root model)

```python
class OkResponse(BaseModel):
    value: int

class ErrorResponse(BaseModel):
    error: str

class MyResponse(BaseModel):
    __root__ = Union[OkResponse, ErrorResponse]

@validate(resp=Response(HTTP_200=MyResponse))
def my_endpoint()
    return {"value": 1}  # Should work (if we want to be consistent with "simple" models
    return {"error": "something bad happened"}  # Should work
```

If we raise a validation error (line 232 in the flask plugin), then we'll fail early and we'll never hit the "second-chance" validation logic (lines 245-251 in the flask plugin). So imo, we should ignore any exception here, and go on.

> BTW, if it's validated by the 1st one successfully, then shall we set the skip_validation to skip the next one?

Are you suggesting to change this (from line 225):

```python
            elif expect_model and is_root_model(expect_model):
                if not isinstance(model, expect_model):
                    # Make it possible to return an instance of the model __root__ type
                    # (i.e. not the root model itself).
                    try:
                        model = expect_model(__root__=model)
                    except ValidationError:
                        pass
                if isinstance(model, expect_model):
                    skip_validation = True
                    result = (model.dict()["__root__"], status, *rest)
```

to this:

```python
            elif expect_model and is_root_model(expect_model):
                if not isinstance(model, expect_model):
                    # Make it possible to return an instance of the model __root__ type
                    # (i.e. not the root model itself).
                    try:
                        model = expect_model(__root__=model)
                        skip_validation = True
                        result = (model.dict()["__root__"], status, *rest)
                    except ValidationError:
                        pass
                elif isinstance(model, expect_model):
                    skip_validation = True
                    result = (model.dict()["__root__"], status, *rest)
```

If so that sounds reasonable!

Reply:
1. For the example, I still don't get it. I feel `Model(__root__=data)` is equivalent to `Model.parse_obj(data)` for the root Model. So if it failed the 1st validation, how can it pass the second one? The difference I can see here is that the 2nd data is already a JSON, if it's a valid JSON, does that mean it should either pass both validations or fail both validations?
2. Yes. One more thing, it might be clearer to use `try...else` here.

Reply:
Another thought on this: can we do the response validation once?

- flask: call the `make_response` after all the validation
- falcon: assign the validated dict to the media: `_resp.media = validated_model.dict()`

Reply:
> I feel Model(__root__=data) is equivalent to Model.parse_obj(data)

Ah very interesting, I assumed pydantic would only accept one of the subtypes defined by `__root__` when initializing a root model with `Model(__root__=data)` (i.e. only `Model(__root__=User(name="user"))` would pass validation, not `Model(__root__={"name": "user1"})`). Turns out all of those work.

```python
class User(BaseModel):
    name: str


class Error(BaseModel):
    message: str


class Response(BaseModel):
    __root__: list[User] | Error


response = Response(__root__=[User(name="hello")])
response = Response(__root__=[{"name": "user1"}])

response = Response(__root__=Error(message="RuntimeError"))
response = Response(__root__={"message": "RuntimeError"})
```

Then that was my bad, and there is indeed no need for the second validation for this to work. I'm gonna rework this a bit.

Reply:
> Another thought on this: can we do the response validation once?

I like the suggestion, but I feel this will lead to larger changes in this PR. I have some ideas how we could make validation simpler/more consistent across plugins and would be happy to contribute additional work towards this. Okay with you to do this in another PR?

Reply:
Sure, we can do this in another PR.